wandb/sdk/internal/internal_api.py

```diff
@@ -887,49 +889,74 @@ def upsert_run(
-            "commit": commit,
-            "displayName": display_name,
-            "notes": notes,
-            "host": None if self.settings().get("anonymous") == "true" else host,
-            "debug": env.is_debug(env=self._environ),
-            "repo": repo,
-            "program": program_path,
-            "jobType": job_type,
-            "state": state,
-            "sweep": sweep_name,
-            "summaryMetrics": summary_metrics,
-        }
+        for i, (r1, r2) in enumerate(zip(replace_1, replace_2)):
+            mutation_filled = mutation_string.replace(
+                "__rewind_replace_1__", r1
+            ).replace("__rewind_replace_2__", r2)
+            mutation = gql(mutation_filled)
 
-        response = self.gql(mutation, variable_values=variable_values, **kwargs)
+            if config is not None:
+                config = json.dumps(config)
+            if not description or description.isspace():
+                description = None
 
-        run = response["upsertBucket"]["bucket"]
-        project = run.get("project")
-        if project:
-            self.set_setting("project", project["name"])
-            entity = project.get("entity")
-            if entity:
-                self.set_setting("entity", entity["name"])
+            kwargs = {}
+            if num_retries is not None:
+                kwargs["num_retries"] = num_retries
 
-        return response["upsertBucket"]["bucket"], response["upsertBucket"]["inserted"]
+            variable_values = {
+                "id": id,
+                "entity": entity or self.settings("entity"),
+                "name": name,
+                "project": project or util.auto_project_name(program_path),
+                "groupName": group,
+                "tags": tags,
+                "description": description,
+                "config": config,
+                "commit": commit,
+                "displayName": display_name,
+                "notes": notes,
+                "host": None if self.settings().get("anonymous") == "true" else host,
+                "debug": env.is_debug(env=self._environ),
+                "repo": repo,
+                "program": program_path,
+                "jobType": job_type,
+                "state": state,
+                "sweep": sweep_name,
+                "summaryMetrics": summary_metrics,
+            }
+
+            if i == 0:
+                variable_values["rewindStep"] = rewind_step
+
+            try:
+                response = self.gql(mutation, variable_values=variable_values, **kwargs)
+            except CommError as e:
+                if "rewind" in e.args[0].lower():
+                    if rewind_step is not None:
+                        raise CommError(
+                            "Run rewinding is not supported on this version of the "
+                            "wandb backend. Please retry your request without rewinding."
+                        )
+                    continue
```

Comment:
This is nasty enough I think it's time to wire through the `serverInfo` version information and do this based on the version of the server...

Reply:
That would be great!

Reply:
If someone wants to do that, it needs to be in the internal process (not the way it is currently implemented), and there are implications in doing that before the first upsert run call.  (new places for network hangs)

Reply:
Given that's going to be a big chore, let's add a big comment saying this needs to be refactored when we pipe it through.

Reply:
@raubitsj I had thought `wandb.login` (which is automatically called by wandb.init) was making a request already that could pipe the data through...

Reply:
nope.  in `wandb.setup()` and it is happening in the user context with a 5 second timeout.
You can not rely on this.